### PR TITLE
feat: add password prompt to `@clack/prompts`

### PR DESCRIPTION
**Pull request by beerose**
I noticed that `@clack/prompts` doesn't export a password prompt, and I needed it in a project I'm currently working on 😅

Let me know if I should make any changes.

### CODE REVIEW

**Comment by ulken** [comment]
Cool beans!🫘 

I haven't looked at the code, nor tested it, yet, but just want to throw a thought out there:

Did you consider just adding the `mask` attribute to the `TextPrompt`?

I thought about adding this myself and was gonna explore that path first. 

Not sure just masking qualifies for a new type of prompt.

Thoughts? Any blockers/features I'm missing?

**Comment by beerose** [comment]
When I used it for the first time, I kind of expected the `text` function to allow overriding how the input is printed (basically a mask), so I guess that's something we could do instead. Since there was a PasswordPromp class in `@clack/core`, I figured I could use it and create a wrapper in `@clack/prompts` (same as, for example, `TextPrompt` is used in `text`).

If we were to add `mask` to `TextPrompt` do you mean:
1) Unifying `TextPrompt` and `PasswordPrompt` classes into a single class.
2) Adding `mask` to `TextPrompt` class and `text` function and leave `PasswordPrompt` as is?
Alternatively we can add `mask` prop to `text` and use either `TextPrompt` or `PasswordPrompt` class?

**Comment by ulken** [comment]
Oh, wow. I had totally missed the password prompt in core. My bad.

Well then, don't mind me. I guess @natemoo-re intended to create a separate component to go with it but hadn't come to it yet (or just missed it).

I was thinking along these lines: https://github.com/vadimdemedes/ink-text-input#mask.

**Comment by ulken** [comment]
Follow-up question: shouldn't `PasswordPrompt` in core extend `TextPrompt` at least? Feels like unnecessary duplication. Or do we expect them to differ more in the future?

File: packages/core/src/prompts/password.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export default class PasswordPrompt extends Prompt {
 	get cursor() {
 		return this._cursor;
 	}
-	private get masked() {
+	get masked() {
```
Comment (ulken) [inline packages/core/src/prompts/password.ts:13]: Unrelated to this PR, but just curious: any particular reason we don't use `replaceAll()` here?

`return this.value.replaceAll(/./g, this._mask);`

Easier to parse/read IMO.

File: packages/prompts/src/index.ts [hunk 2]
```diff
@@ -87,6 +88,38 @@ export const text = (opts: TextOptions) => {
...
+			switch (this.state) {
+				case 'error':
+					return `${title.trim()}\n${color.yellow(S_BAR)}  ${masked}\n${color.yellow(
+						S_BAR_END
+					)}  ${color.yellow(this.error)}\n`;
+				case 'submit':
+					return `${title}${color.gray(S_BAR)}  ${color.dim(masked)}`;
+				case 'cancel':
+					return `${title}${color.gray(S_BAR)}  ${color.strikethrough(color.dim(masked ?? ''))}${
+						masked?.trim() ? '\n' + color.gray(S_BAR) : ''
```
Comment (ulken) [inline packages/prompts/src/index.ts:114]: Haven't tested, but is `trim()` legit in this case? A space would be a valid password character, no?

Comment (ulken) [reply packages/prompts/src/index.ts:114]: Or rather, would this ever trim anything as spaces would be replaced by the masking character?

Comment (beerose) [reply packages/prompts/src/index.ts:114]: I followed the implementation of the `text` prompt (https://github.com/natemoo-re/clack/blob/main/packages/prompts/src/index.ts#L74) here, but I guess a space at the end is usually trimmed when you provide a password? Anyway, it won't change the underlying value, as this part of the code only prints stuff to the user's console. But let me know if I should remove the `trim()`.

File: packages/core/src/prompts/password.ts (hunk 1 above)
Comment (ulken) [reply packages/core/src/prompts/password.ts:13]: Yes, sorry. They were intended for him. Should have @mentioned him.

File: packages/prompts/src/index.ts (hunk 2 above)
Comment (ulken) [reply packages/prompts/src/index.ts:114]: > but I guess a space at the end is usually trimmed when you provide a password

I think not. I'd say that's bad practice.

> Anyway, it won't change the underlying value, as this part of the code only prints stuff to the user's console.

Yes, but what I'm saying is will this ever actually change anything? If not, I'd argue we remove it. If spaces are replaced with the mask character there will never be anything to trim.

File: packages/core/src/prompts/password.ts (hunk 1 above)
Comment (natemoo-re) [reply packages/core/src/prompts/password.ts:13]: No particular reason! I'm cool with that change.

File: packages/prompts/src/index.ts (hunk 2 above)
Comment (natemoo-re) [reply packages/prompts/src/index.ts:114]: Agreed, the trim can be removed in this case since passwords should be passed literally

**Comment by natemoo-re** [comment]
> Follow-up question: shouldn't `PasswordPrompt` in core extend `TextPrompt` at least? Feels like quite a natural extension and unnecessary duplication. Or do we expect them to differ more in the future?

That seems like a very reasonable way to do it. If you want to PR that as a refactor, go for it! I'm all for anything that reduces copy + paste.

**Comment by ulken** [comment]
All right, I'll do the other changes in a separate PR. So this should be good to go then?

**Review by natemoo-re — APPROVED** [review]
🎉 Thanks for the PR!

**Pull request merged**